foundations/foundation-vertx/src/main/java/org/apache/servicecomb/foundation/vertx/stream/OutputStreamToWriteStream.java

```diff
@@ -0,0 +1,200 @@
+
+import org.apache.servicecomb.foundation.common.io.AsyncCloseable;
+
+import io.vertx.core.Context;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.streams.WriteStream;
+
+/**
+ * for pump from a readStream
+ */
+public class OutputStreamToWriteStream implements WriteStream<Buffer>, AsyncCloseable<Void> {
+  private static final int DEFAULT_MAX_BUFFERS = 4;
+
+  private static final int SMALLEST_MAX_BUFFERS = 2;
+
+  private OutputStream outputStream;
+
+  private Context context;
+
+  private boolean autoCloseOutputStream;
+
+  private Handler<Throwable> exceptionHandler;
+
+  // resume readStream
+  private Handler<Void> drainHandler;
+
+  // when invoke close, but outputStream not write all data, must put close logic to closedDeferred
+  private Runnable closedDeferred;
+
+  private boolean closed;
+
+  // buffers.size() need to loop all node, and maybe result is not correct in concurrent condition
+  // we just need to flow control by pump, so use another size
+  private Queue<Buffer> buffers = new ConcurrentLinkedQueue<>();
+
+  private int currentBufferCount;
+
+  // just indicate if buffers is full, not control add logic
+  // must >= SMALLEST_MAX_BUFFERS
+  // if < SMALLEST_MAX_BUFFERS, then maxBuffers will be SMALLEST_MAX_BUFFERS
+  private int maxBuffers = DEFAULT_MAX_BUFFERS;
+
+  // if currentBufferCount <= drainMark, will invoke drainHandler to resume readStream
+  private int drainMark = maxBuffers / 2;
+
+  public OutputStreamToWriteStream(Context context, OutputStream outputStream,
+      boolean autoCloseOutputStream) {
+    this.context = context;
+    this.outputStream = outputStream;
+    this.autoCloseOutputStream = autoCloseOutputStream;
+  }
+
+  @Override
+  public WriteStream<Buffer> exceptionHandler(Handler<Throwable> handler) {
+    this.exceptionHandler = handler;
+    return this;
+  }
+
+  private void handleException(Throwable t) {
+    if (exceptionHandler != null) {
+      exceptionHandler.handle(t);
+    }
+  }
+
+  @Override
+  public synchronized WriteStream<Buffer> write(Buffer data) {
+    currentBufferCount++;
+    buffers.add(data);
+    context.executeBlocking(this::writeInWorker,
+        true,
+        ar -> {
+          if (ar.failed()) {
+            handleException(ar.cause());
+          }
+        });
+    return this;
+  }
+
+  protected void writeInWorker(Future<Object> future) {
+    while (true) {
+      Buffer buffer = buffers.poll();
+      if (buffer == null) {
+        future.complete();
+        return;
+      }
+
+      try {
+        outputStream.write(buffer.getBytes());
+
+        synchronized (OutputStreamToWriteStream.this) {
+          currentBufferCount--;
+          Runnable action = (currentBufferCount == 0 && closedDeferred != null) ? closedDeferred : this::checkDrained;
+          action.run();
+        }
+      } catch (IOException e) {
+        currentBufferCount--;
```

Comment:
Here is no sychronized.   If we just need to make sure currentBufferCount is automatic, we don't need to use the synchroized to do this job.

Reply:
not just protect currentBufferCount, we must protect close/drain logic.